pkg/security/ebpf/probes/rawpacket/pcap.go

```diff
@@ -110,40 +124,75 @@ func BPFFilterToInsts(index int, filter string, opts ProgOpts) (asm.Instructions
 
 	// add nop insts, used to test the max insts and artificially generate tail calls
 	for i := 0; i != opts.NopInstLen; i++ {
+		// insert a nop instruction
 		insts = append(insts,
-			asm.JEq.Imm(asm.R9, 0, opts.sendEventLabel).WithSymbol(resultLabel),
+			asm.JEq.Imm(opts.ctxSaveReg, 0, opts.onMatchLabel).WithSymbol(resultLabel),
 		)
 		resultLabel = ""
 	}
 
-	// filter result
-	insts = append(insts,
-		asm.JNE.Imm(cbpfcOpts.Result, 0, opts.sendEventLabel).WithSymbol(resultLabel),
-	)
+	mismatchLabel := fmt.Sprintf("mismatch_%d_", index)
 
+	if filter.Pid != 0 {
+		insts = append(insts,
+			// == 0, no match
+			asm.JEq.Imm(cbpfcOpts.Result, 0, mismatchLabel).WithSymbol(resultLabel),
+
+			// check the pid
+			// load the pid from the packet
+			asm.LoadMem(asm.R7, opts.eventPtrReg, structRawPacketEventPidOffset, asm.Word),
+			asm.JEq.Imm(asm.R7, int32(filter.Pid), opts.onMatchLabel),
+			asm.Mov.Imm(asm.R4, 0).WithSymbol(mismatchLabel), // nop instruction, just hold the symbol
+		)
+	} else if !filter.CGroupPathKey.IsNull() {
```

Comment:
Should we also check if `hasGetCurrentCgroupId` is true here?

Reply:
I don't think this is required as we use a fallback kernel side when this helper is not present. [pkg/security/ebpf/probes/rawpacket/pcap.go](https://github.com/DataDog/datadog-agent/pull/39066/files/d91e74ee6e6c7a68834c47dc8c9e471aeeb98774#diff-81853b5953ffa7c874e254ecb233a0eb262dd713e2d266c6c0b143277bc04b69)